### ISSUE TITLE
Cloudflare v10 causes wrong paths in generated `_routes.json`

### ISSUE THREAD

**Issue by dallyh** [withastro/adapters#216]
### Astro Info

```block
Astro                    v4.5.12
Node                     v20.12.0
System                   Windows (x64)
Package Manager          npm
Output                   hybrid
Adapter                  @astrojs/cloudflare
Integrations             @astrojs/react
```

### Describe the Bug

After upgrade to v10 of the Cloudflare adapter, some routes in my project became unavailable. I investigated and found out, that the `_routes.json` differs from previous version (9.2.0).

Routes from 10.0.1:

```json
{
  "version": 1,
  "include": [
    "/*"
  ],
  "exclude": [
    "/*"
  ]
}
```

Routes from 9.2.0:

```json
{
  "version": 1,
  "include": [
    "/",
    "/_image",
    "/*/contact"
  ],
  "exclude": []
}
```

This of course causes the server endpoint "/" to be not found, as well as other server endpoints.

Steps to reproduce
- pnpm install
- pnpm build
- Observe `_routes.json` in `/dist`
- Run pnpm run preview, server endpoint `/` causes 404
- Change `@astrojs/cloudflare` version in packages to `9.2.0`
- pnpm install => pnpm build
- Observe  `_routes.json` in `/dist`
- Run pnpm run preview, server endpoint `/` causes redirect to `/contact` as it should

Also the route for `/contact` which is server rendered should have (in my opinion) two routes in the `routes.json`, and that is `/contact` and `/*/contact`, because the REST parameter `[...lang]` can be undefined. This is how Astro's `dev` server behaves. When running this in preview mode with Wrangler, then `/contact` is not found and `//contact` is found (notice the empty `//`), however Astro's `dev` server finds both of the routes. 

In the example use the branch `cf-10-routes-repro`.

### What's the expected result?

Routes should be generated accordingly to all of the SSR/Static paths.

### Link to Minimal Reproducible Example

https://github.com/dallyh/daliborhon.dev/tree/cf-10-routes-repro

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by alexanderniebuhr** [comment]
Thank you, actually this is a good edge case, fix will come soon!

**Comment by twodft** [comment]
Hi, also to mention that `CLOUDFLARE_COMBINED_LIMIT` seems to have a problem too, got error when running wrangler preview. 
```
✘ [ERROR] Invalid _routes.json file found at: D:\projects\website\dist

  Detected rules that are over the 100 rule limit. Please make sure you have a total of 100 include
  and exclude rules combined.
```

in my case `_routes.json` has a combined 104 include/exclude rules, extend rules seems not take into account
```
output: 'hybrid',
	adapter: cloudflare({
		routes: {
			extend: {
				include: [
					{ pattern: '/api/*' },
					{ pattern: '/_test/*' },
					{ pattern: '/proxy*' },
					{ pattern: '/search*' },
				],
				exclude: [{ pattern: '/*' }],
			},
		},
		platformProxy: {
			enabled: true,
		},
	}),
```

**Comment by alexanderniebuhr** [comment]
@twodft Thank you for letting me know, that information was very helpful. I fixed the logic for checking the length, so it shouldn't occur anymore. However I just want to let you know that your extend patterns don't really make sense to me and will end up in another error, which states that you have duplicated route definitions. Which is expected given the extend patterns in your message.

**Comment by dallyh** [comment]
@alexanderniebuhr I've just tried it out and I'm not entirely sure if it's okay. The routes generated now look like this:

```json
{
  "version": 1,
  "include": [
    "/*"
  ],
  "exclude": [
    "/404",
    "/en/404",
    "/en/blog/posts/01",
    "/en/blog/posts/02",
    "/en/blog/posts/03",
    "/en/blog/posts/04",
    "/en/blog/posts/05",
    "/en/blog/posts/06",
    "/en/blog/posts/07",
    "/blog/posts/01",
    "/blog/posts/02",
    "/blog/posts/03",
    "/blog/posts/04",
    "/blog/posts/05",
    "/blog/posts/06",
    "/blog/posts/07",
    "/CNAME",
    "/favicon.ico",
    "/_astro/*"
  ]
}
```

Imagine having a blog with 100+ posts, then we wuld reach the limit of Cloudflare's `_routes.json`.

What happened that the routes looked fine in the version 9.2.0 and now they are entirely different? I'm convinced hat before they were generated better, as every server endpoin was in the include, and all static endpoints were not written into the paths. This approach was miles better than the current one, as it was less likely to hit the 100 limit.

```json
{
  "version": 1,
  "include": [
    "/",
    "/_image",
    "/*/contact"
  ],
  "exclude": []
}
```

If it has to stay like this, then may I propose at least that the routes looks like this in those cases?

```json
{
  "version": 1,
  "include": [
    "/*"
  ],
  "exclude": [
    "/404",
    "/en/404",
    "/en/blog/posts/*",
    "/blog/posts/*",
    "/CNAME",
    "/favicon.ico",
    "/_astro/*"
  ]
}
```

Thanks!

**Comment by alexanderniebuhr** [comment]
> then we would reach the limit of Cloudflare's _routes.json

You will not, there won't be added more routes than the limit.

> I'm convinced that before they were generated better

That's not correct. The previous generated routes of v9, wouldn't work, because we can't automatically discover a prerended 404 at the root of your project, and for projects with a on-demand rendered 404, we need to include everything with `/*`

There is no way to generate the second proposal automatically, because in theory there could be a on-demand rendered page under `/blog/posts/*`

**Comment by alexanderniebuhr** [comment]
Ok I improved the 404 detection logic, which should also pick up prerendered 404 pages if they are nested under a dynamic segment, which can be undefined.

> The structure is similiar to the provided example

That is not true. In your screenshot you see that the api server endpoints are directly under `pages`, while in the other example the contact ondemand rendered pages are nested under the dynamic segment `[...lang]`.. That leads to different outputs. I've updated the generation that the `_routes.json` is similar to the v9 version for the file tree of the screenshot.. For the case of `/[...lang]/contact`, we can't get it generated like that automatically. 

> Also it is just that the behavior changed a lot from a version to version, and from my point of view the update broke a lot of things, so please bear with me on this.

I'm sorry that it broke things for you. That's the reason why it is a major version bump _(according to semver with breaking changes)_ and why we still have maintenance support for v9 for at least the next 4 weeks.

You can try this version to see if it is closer to your preferred output: `@astrojs/cloudflare@0.0.0-cf-10-routes-fix-20240330133117`

**Comment by dallyh** [comment]
@alexanderniebuhr Oh my bad, i missed the ondemand contact page in the example. From the last patch everything looks great! Thank you very much.

Also don't be sorry for the broken things, it is a major bump, so things like this are expected even from your side. I can say that I like the things that v10 does better than v9, it is indeed more streamlined. So yeah thanks!
